File: app/src/main/java/com/mystery0/hitokoto/widget/HitokotoWidget.java
package com.mystery0.hitokoto.widget;

import android.Manifest;
import android.app.PendingIntent;
import android.appwidget.AppWidgetManager;
import android.appwidget.AppWidgetProvider;
import android.content.Context;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.graphics.Color;
import android.support.v4.content.ContextCompat;
import android.util.TypedValue;
import android.view.View;
import android.widget.RemoteViews;
import android.widget.Toast;

import com.mystery0.hitokoto.App;
import com.mystery0.hitokoto.class_class.Hitokoto;
import com.mystery0.hitokoto.R;
import com.mystery0.hitokoto.SettingsActivity;
import com.mystery0.tools.Logs.Logs;

public class HitokotoWidget extends AppWidgetProvider
{
	private static final String TAG = "HitokotoWidget";

	static void updateAllAppWidget(String text, String source)
	{
		Context context = App.getContext();
		AppWidgetManager appWidgetManager = AppWidgetManager.getInstance(context);
		for (int id : App.getIdsSet())
		{
			RemoteViews remoteViews;
			switch (App.getWidgetConfigure().getTextAligned())
			{
				case 0:
					remoteViews = new RemoteViews(context.getPackageName(), R.layout.hitokoto_widget_left);
					break;
				case 2:
					remoteViews = new RemoteViews(context.getPackageName(), R.layout.hitokoto_widget_right);
					break;
				default:
					remoteViews = new RemoteViews(context.getPackageName(), R.layout.hitokoto_widget_center);
					break;
			}
			remoteViews.setViewPadding(R.id.appwidget_source, 0, App.getWidgetConfigure().getTextPadding(), 0, 0);
			remoteViews.setTextViewText(R.id.appwidget_text, text);
			remoteViews.setTextViewText(R.id.appwidget_source, source);
			remoteViews.setTextColor(R.id.appwidget_text, Color.parseColor(App.getWidgetConfigure().getTextColor()));
			remoteViews.setTextColor(R.id.appwidget_source, Color.parseColor(App.getWidgetConfigure().getTextColor()));
			if (App.getWidgetConfigure().getNotShowSource())
			{
				remoteViews.setViewVisibility(R.id.appwidget_source, View.GONE);
			} else
			{
				remoteViews.setViewVisibility(R.id.appwidget_source, View.VISIBLE);
			}
			remoteViews.setTextViewTextSize(R.id.appwidget_text, TypedValue.COMPLEX_UNIT_SP, App.getWidgetConfigure().getTextSize());
			remoteViews.setTextViewTextSize(R.id.appwidget_source, TypedValue.COMPLEX_UNIT_SP, App.getWidgetConfigure().getTextSize());
			if (App.getWidgetConfigure().getClickToRefresh())
			{
				Intent topIntent = new Intent(context, OnClickService.class);
				PendingIntent topPendingIntent = PendingIntent.getService(context, 0, topIntent, PendingIntent.FLAG_UPDATE_CURRENT);
				remoteViews.setOnClickPendingIntent(R.id.appwidget_text, topPendingIntent);
				remoteViews.setOnClickPendingIntent(R.id.appwidget_source, topPendingIntent);
			}
			appWidgetManager.updateAppWidget(id, remoteViews);
		}
	}

	private void initWidget(Context context, int appId, AppWidgetManager appWidgetManager)
	{
		RemoteViews remoteViews;
		String[] strings = App.getWidgetConfigure().getTemp();
		switch (App.getWidgetConfigure().getTextAligned())
		{
			case 0:
				remoteViews = new RemoteViews(context.getPackageName(), R.layout.hitokoto_widget_left);
				break;
			case 2:
				remoteViews = new RemoteViews(context.getPackageName(), R.layout.hitokoto_widget_right);
				break;
			default:
				remoteViews = new RemoteViews(context.getPackageName(), R.layout.hitokoto_widget_center);
				break;
		}
		String[] texts = App.getWidgetConfigure().getTemp();
		remoteViews.setViewPadding(R.id.appwidget_source, 0, App.getWidgetConfigure().getTextPadding(), 0, 0);
		remoteViews.setTextViewText(R.id.appwidget_text, texts[0]);
		remoteViews.setTextViewText(R.id.appwidget_source, texts[1]);
		remoteViews.setTextColor(R.id.appwidget_text, Color.parseColor(App.getWidgetConfigure().getTextColor()));
		remoteViews.setTextColor(R.id.appwidget_source, Color.parseColor(App.getWidgetConfigure().getTextColor()));
		if (App.getWidgetConfigure().getNotShowSource())
		{
			remoteViews.setViewVisibility(R.id.appwidget_source, View.GONE);
		} else
		{
			remoteViews.setViewVisibility(R.id.appwidget_source, View.VISIBLE);
		}
		if (App.getWidgetConfigure().getClickToRefresh())
		{
			Intent topIntent = new Intent(context, OnClickService.class);
			PendingIntent topPendingIntent = PendingIntent.getService(context, 0, topIntent, PendingIntent.FLAG_UPDATE_CURRENT);
			remoteViews.setOnClickPendingIntent(R.id.appwidget_text, topPendingIntent);
			remoteViews.setOnClickPendingIntent(R.id.appwidget_source, topPendingIntent);
		}
		remoteViews.setTextViewTextSize(R.id.appwidget_text, TypedValue.COMPLEX_UNIT_SP, App.getWidgetConfigure().getTextSize());
		remoteViews.setTextViewTextSize(R.id.appwidget_source, TypedValue.COMPLEX_UNIT_SP, App.getWidgetConfigure().getTextSize());
		appWidgetManager.updateAppWidget(appId, remoteViews);
		if (strings[1].equals("开发者"))
		{
			App.getWidgetConfigure().refreshText();
		}
	}

	@Override
	public void onUpdate(Context context, AppWidgetManager appWidgetManager, int[] appWidgetIds)
	{
		Logs.i(TAG, "onUpdate: ");
		for (int id : appWidgetIds)
		{
			App.addID(id);
			initWidget(context, id, appWidgetManager);
		}
	}

	@Override
	public void onDeleted(Context context, int[] appWidgetIds)
	{
		Logs.i(TAG, "onDeleted: ");
		for (int id : appWidgetIds)
		{
			App.removeID(id);
		}
	}

	@Override
	public void onEnabled(Context context)
	{
		Logs.i(TAG, "onEnabled: ");
		App.getWidgetConfigure().setEnable(true);
		if (ContextCompat.checkSelfPermission(context, Manifest.permission.WRITE_EXTERNAL_STORAGE)
				!= PackageManager.PERMISSION_GRANTED)
		{
			context.startActivity(new Intent(context, SettingsActivity.class));
			Toast.makeText(context, R.string.hint_permission, Toast.LENGTH_SHORT)
					.show();
		}
		Intent intent = new Intent(context, WidgetService.class);
		context.startService(intent);
	}

	@Override
	public void onDisabled(Context context)
	{
		App.getWidgetConfigure().setEnable(false);
		Logs.i(TAG, "onDisabled: ");
		Intent intent = new Intent(context, WidgetService.class);
		context.stopService(intent);
	}

	@Override
	public void onReceive(Context context, Intent intent)
	{
		super.onReceive(context, intent);
		String action = intent.getAction();
		Logs.i(TAG, "onReceive: " + action);
		if ("android.appwidget.action.APPWIDGET_UPDATE".equals(action))
		{
			Hitokoto hitokoto = (Hitokoto) intent.getSerializableExtra(context.getString(R.string.hitokoto_object));
			if (hitokoto != null)
			{
				Logs.i(TAG, "onReceive: " + hitokoto.getHitokoto());
				Logs.i(TAG, "onReceive: " + hitokoto.getFrom());
				updateAllAppWidget(hitokoto.getHitokoto(), "————" + hitokoto.getFrom());
			}
		}
	}
}

File: app/src/main/java/com/mystery0/hitokoto/widget/HitokotoWidget.kt
package com.mystery0.hitokoto.widget

import android.app.PendingIntent
import android.appwidget.AppWidgetManager
import android.appwidget.AppWidgetProvider
import android.content.Context
import android.content.Intent
import android.widget.RemoteViews

import com.mystery0.hitokoto.R
import com.mystery0.hitokoto.activity.HitokotoWidgetConfigureActivity
import com.mystery0.hitokoto.classes.Hitokoto
import com.mystery0.hitokoto.constant.IntentConstant
import com.mystery0.hitokoto.service.HitokotoIntentService
import vip.mystery0.logs.Logs

/**
 * Implementation of App Widget functionality.
 * App Widget Configuration implemented in [HitokotoWidgetConfigureActivity]
 */
class HitokotoWidget : AppWidgetProvider() {

	override fun onUpdate(context: Context, appWidgetManager: AppWidgetManager, appWidgetIds: IntArray) {
		val stringBuilder = StringBuilder()
		appWidgetIds.forEach {
			stringBuilder.append(it).append(" ")
			list.add(it)
		}
		context.startService(Intent(context, HitokotoIntentService::class.java))
		Logs.i("onUpdate: $stringBuilder")
	}

	override fun onDeleted(context: Context, appWidgetIds: IntArray) {
		val stringBuilder = StringBuilder()
		appWidgetIds.forEach {
			stringBuilder.append(it).append(" ")
			list.remove(it)
		}
		Logs.i("onDeleted: ")
	}

	override fun onEnabled(context: Context) {
		Logs.i("onEnabled: ")
	}

	override fun onDisabled(context: Context) {
		Logs.i("onDisabled: ")
	}

	override fun onReceive(context: Context, intent: Intent) {
		Logs.i("onReceive: ${intent.action}")
		if (intent.action == "android.appwidget.action.APPWIDGET_UPDATE") {
			val hitokoto = intent.getParcelableExtra<Hitokoto>(IntentConstant.intent_hitokoto)
			if (hitokoto != null) {
				Logs.i("onReceive: $hitokoto")
				updateAppWidget(context, hitokoto)
			}
		}
	}

	companion object {
		private val list = ArrayList<Int>()

		internal fun updateAppWidget(context: Context, hitokoto: Hitokoto) {
			val views = RemoteViews(context.packageName, R.layout.hitokoto_widget)
			val appWidgetManager = AppWidgetManager.getInstance(context)
			val widgetText = "${hitokoto.hitokoto}\n——${hitokoto.from}"
			views.setTextViewText(R.id.appwidget_text, widgetText)
			list.forEach {
				appWidgetManager.updateAppWidget(it, views)
			}
			val intent = Intent(context, HitokotoIntentService::class.java)
			views.setOnClickPendingIntent(R.id.appwidget_text, PendingIntent.getService(context, 0, intent, PendingIntent.FLAG_UPDATE_CURRENT))
		}
	}
}

